### PR TITLE
WIP: Alternative architecture

### PR DESCRIPTION
**Pull request by j-maas**
I have an idea for improving the architecture. You almost already have it in the echobot example, however, there seem to be a few things that go wrong somewhere. Unfortunately, I don't yet understand these kinds of conflicts well, but I think the idea still comes across.

Basically, I would like to decouple the update handling from the update loop. Because the bots only really differ in the handler, not so much in the runtime/loop. The update handler would simply receive each update, and simply construct all the requests it wants to send in response. Very simple, functional architecture. (For returning multiple requests, even with a delay between them, the handler could return a stream. For now, I think simply returning a list of requests is enough, and it should be simple enough to upgrade when we need to.)

I tried to do that in the echobot example. However, there are some trait issues that I don't fully understand. My guess is that `ApiRequest` being a struct does not mix well with the generic `Request` trait, because then the `TResponse` type cannot be inferred for `ApiRequest`. Maybe making `ApiRequest` a trait as well would help with this, I'm not sure.

But before I dive deeper into this issue, I wanted to ask you what you think of this general idea?

### CODE REVIEW

**Comment by Leonqn** [comment]
I think that `handle_update` function is to simplified. Often after receiving message you need to make rpc call, read from disk, etc and this signature will be to restrictive. For example i have bot that coverts voice messages to text and [for one message it make several rpc calls](https://github.com/Leonqn/speech-to-text-bot/blob/master/recognition_bot/src/bot.rs#L313), but for [other it makes no calls](https://github.com/Leonqn/speech-to-text-bot/blob/master/recognition_bot/src/bot.rs#L288). Maybe we should make the signature like this: `Update -> impl IntoFuture<Item=Option<impl Request>, Error=()> ? ` at the beginning?


Also i think that we should keep raw updates stream accessible from client, however i like this idea and it may be an extension to our stream.

**Comment by j-maas** [comment]
I completely agree! The reason I used this simple signature for now is that it works for the example. It can't even send multiple replies, so I know it's really limited. I fully agree with your proposal to make it asynchronous.

I also agree that the runtime/loop should stay accessible. My suggestion would be to offer some predefined "runtimes", one for polling, one for webhooks, if the user doesn't care. But everything these systems use should be available to the user, such that they can modify it or make a custom runtime.

In summary, we would architecture it such that we can cut the update handler from the update loop, which right now does not seem easily possible.

**Comment by Leonqn** [comment]
I've fixed old compilation issues, but add another one :)

**Comment by j-maas** [comment]
Unfortunately, I'm not an expert on this part of Rust, yet. But to me it looks like the request types should own all their data. So no `&str`, only `String`, and for all other things as well. That seems to work.

I think the reason you might not want to let every request own its data is that things like chat ids might be shared across multiple requests, and thus this might be inefficient. But it seems to work at least, and for now I can't figure out how to make it work with references. Do you know how that would work?

**Comment by j-maas** [comment]
Ok, I think the original issue might have to do with dynamic dispatch in Rust: https://doc.rust-lang.org/book/ch17-02-trait-objects.html#using-trait-objects-that-allow-for-values-of-different-types

They need to be behind a pointer, because Rust needs to know the size of variables at compile time, which by design is not the case for dynamic dispatch.

However, it seems to work now. I'm still not entirely sure about whether there is a better way than to make each `Request` own all its data, but I think that's something we can figure out later.

**Comment by Leonqn** [comment]
I don't like the idea of owned values in Request structs. 
I made all requests containing references because it is easy put to them static strings as well as other references without uneccessary allocations.

We have ApiRequest struct that represents ready request and it contains only owned values, maybe we can do something with this? Or create another struct that contains only serialized request body and `handle` function will return that struct?

**Comment by j-maas** [comment]
I'm not that good at Rust, so I often have trouble with making references work. But if in the end the `ApiRequest` owns everything anyway, it feels like `Request` could as well.

Just for me to understand better, is there an example where you need `Request` to have references that doesn't work as well when it owns its data?

**Comment by Leonqn** [comment]
In echobot example you should call .to_owned() on static string and allocate string on heap. ApiRequest owns serialized data that can be copied from references. 
We actually don't need to own data for sending request because we will copy it anyway. And i think that it is better to not own data if it is possible.

**Comment by j-maas** [comment]
I guess I still have a lot to learn... :D

One problem I see is that `Request` and `ApiRequest` are generic over the response type, which does not mix well with my approach to return arbitrary `Request`s, because of issues with Rust's handling of dynamic dispatch. Maybe we can add some mechanism that "consumes" (or ignores) the response, such that we can have some non-generic type that works with dynamic dispatch...

I hope this makes sense, because I'm still trying to figure it out myself. 😅

**Comment by Leonqn** [comment]
>Maybe we can add some mechanism that "consumes" (or ignores) the response, such that we can have some non-generic type that works with dynamic dispatch...

Yes, i think we can.

Also there is one more with this `handle` function. What if request sending failed? How can we tell user that he needs to retry it or something like that?

**Comment by j-maas** [comment]
That's an interesting question. In most cases, failed requests would be due to the request violating some conditions (text too long, invalid id). We can enforce these conditions in our types, so that should be taken care of.

In case the network is down, the "runtime" could store the failed messages, and retry later. No need to think about this in the handler.

But I think there might be some requests, where the response is interesting. In these cases we might need to add such a response mechanism. I imagine that mechanism to be just like Future, where you could simply `.ignore()` or `on_response(|| do_stuff())`. To prevent boilerplate, I could even imagine a macro `requests![first, second, third]` that would automatically `.ignore()` everything that hasn't been handled explicitely.

But these are just ideas. I'd have to tinker a bit to see what works. Right now I want to spend time on other things, but I'm pretty sure I will come back to this in a while. ;)

**Comment by j-maas** [comment]
I just found [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html), which could help with the ownership problems in `Request`. You could give it a `&str`, preventing allocation, but you could also pass it a `String` to own. That resolves the issue that if we move `Update` into the handler, all its values would be dropped and references would be therefore invalid.

What do you think of `Cow`? Should we switch all usage of `&'a str` to `Cow<'a, str>`?

**Comment by Leonqn** [comment]
Cow comes with a lot of boilerplate, i don't like it in this case :) 
Also it just hides allocations when you put a reference

**Comment by j-maas** [comment]
I've tried for a few hours to make this somehow work with trait objects. I failed. It always wants to know the associated types, and I can't figure out, how to get rid of them. As long as I can't get rid of them, I won't be able to return an abstract `Request` from `handle_update`...

I really hope I'm just missing something. The only other way I see to get around that issue is to use enums for the `Request`s instead of traits. That way, there is also no dynamic dispatch. But I want to make it work as is...

**Comment by Leonqn** [comment]
>The only other way I see to get around that issue is to use enums for the Requests instead of traits

With enums we still need to keep somethere return type of sent message. 
Have you tried with assoc type Box<DeserializeOwned + Send + Sync>?

**Comment by j-maas** [comment]
I am still struggling to make this work. :D I think I will do some experiments by starting from scratch. When I have found something that works, I will come back and present it to you.

**Comment by j-maas** [comment]
As you might have noticed, I'm currently concentrating on other projects. I hope I will find the time to come back, since your approach is really interesting!

Keep up the good work!
